File: src/2025-02-23-deepclaude-try/java/exercises/practice/dominoes/src/main/java/Dominoes.java
import java.util.List;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

class Dominoes {

    List<Domino> formChain(List<Domino> inputDominoes) throws ChainNotFoundException {
        if (inputDominoes.isEmpty()) {
            return new ArrayList<>();
        }

        boolean[] used = new boolean[inputDominoes.size()];
        List<Domino> chain = new ArrayList<>();
        List<Domino> result = backtrack(inputDominoes, used, chain, -1);
        if (result != null) {
            return result;
        } else {
            throw new ChainNotFoundException();
        }
    }

    private List<Domino> backtrack(List<Domino> dominoes, boolean[] used, List<Domino> chain, int currentEnd) {
        if (chain.size() == dominoes.size()) {
            if (chain.get(0).getLeft() == chain.get(chain.size() - 1).getRight()) {
                return new ArrayList<>(chain);
            } else {
                return null;
            }
        }

        for (int i = 0; i < dominoes.size(); i++) {
            if (!used[i]) {
                Domino original = dominoes.get(i);
                // Try original orientation
                if (chain.isEmpty() || original.getLeft() == currentEnd) {
                    used[i] = true;
                    chain.add(original);
                    List<Domino> result = backtrack(dominoes, used, chain, original.getRight());
                    if (result != null) {
                        return result;
                    }
                    chain.remove(chain.size() - 1);
                    used[i] = false;
                }
                // Try reversed orientation
                Domino reversed = new Domino(original.getRight(), original.getLeft());
                if (chain.isEmpty() || reversed.getLeft() == currentEnd) {
                    used[i] = true;
                    chain.add(reversed);
                    List<Domino> result = backtrack(dominoes, used, chain, reversed.getRight());
                    if (result != null) {
                        return result;
                    }
                    chain.remove(chain.size() - 1);
                    used[i] = false;
                }
            }
        }
        return null;
    }

}

File: src/2025-02-23-deepclaude-try/python/exercises/practice/dominoes/dominoes.py
def can_chain(dominoes):
    if not dominoes:
        return []
    
    n = len(dominoes)
    if n == 1:
        a, b = dominoes[0]
        return dominoes if a == b else None

    def backtrack(chain, used, current_end):
        if len(chain) == n:
            return chain[0][0] == current_end
        for i in range(n):
            if not used[i]:
                domino = dominoes[i]
                for a, b in [domino, (domino[1], domino[0])]:
                    if a == current_end:
                        used[i] = True
                        chain.append((a, b))
                        if backtrack(chain, used, b):
                            return True
                        chain.pop()
                        used[i] = False
        return False

    for start_idx in range(n):
        for first in [dominoes[start_idx], (dominoes[start_idx][1], dominoes[start_idx][0])]:
            used = [False] * n
            used[start_idx] = True
            chain = [first]
            if backtrack(chain, used, first[1]):
                return chain
    return None
